src/wh_nvm_flash.c

```diff
@@ -440,6 +449,33 @@ static int nfPartition_ProgramInit(whNvmFlashContext* context, int partition)
     return ret;
 }
 
+/*
+ * Checks that the range of units specified by units_offset + units_count
+ * is located inside the specified partition
+ */
+static int nfPartition_CheckRangeUnits(whNvmFlashContext* context,
+                                       int partition,
+                                       uint32_t units_offset,
+                                       uint32_t units_count)
+{
+    size_t maxUnits;
+    size_t partDataBase;
+
+    partDataBase = nfPartition_DataOffset(context, partition);
+    maxUnits = nfPartition_Offset(context, partition)
+               + context->partition_units;
+
+    if (units_offset < partDataBase) {
+        return WH_ERROR_BADARGS;
+    }
+
+    if (units_offset + units_count > maxUnits) {
+        return WH_ERROR_BADARGS;
+    }
+
+    return WH_ERROR_OK;
+}
```

Comment:
Hrmm..  I think this function should operate in bytes instead of units.  Imagine trying to handle the edge case of accessing the last byte in the partition.  Also, the name of this function should include Data, since the starting offset is specific to the data portion of the partition.  This change also corrects the math error below.

```suggestion
/*
 * Checks that the range of bytes specified by byte_offset +  byte_count
 * is located inside the specified partition's data area
 */
static int nfPartition_CheckDataRange(whNvmFlashContext* context,
                                       int partition,
                                       uint32_t byte_offset,
                                       uint32_t byte_count)
{
    uint32_t maxOffset;
    uint32_t partDataBase;

    partDataBase = nfPartition_DataOffset(context, partition) * WHFU_BYTES_PER_UNIT;
    maxOffset = (nfPartition_Offset(context, partition)
               + context->partition_units) * WHFU_BYTES_PER_UNIT;

    if (byte_offset < partDataBase) {
        return WH_ERROR_BADARGS;
    }

    if (byte_offset + byte_count > maxOffset) {
        return WH_ERROR_BADARGS;
    }

    return WH_ERROR_OK;
}
```